include/gridtools/reduction/gpu.hpp

```diff
@@ -0,0 +1,324 @@
+                val = f(val, __shfl_down(val, 4));
+                val = f(val, __shfl_down(val, 2));
+                return f(val, __shfl_down(val, 1));
+#else
+                val = f(val, __shfl_down_sync(mask, val, 16));
+                val = f(val, __shfl_down_sync(mask, val, 8));
+                val = f(val, __shfl_down_sync(mask, val, 4));
+                val = f(val, __shfl_down_sync(mask, val, 2));
+                return f(val, __shfl_down_sync(mask, val, 1));
+#endif
+            }
+
+#if __CUDA_ARCH__ >= 800
+            GT_FUNCTION_DEVICE auto warp_reduce(plus, unsigned mask, int res) { return __reduce_add_sync(mask, res); }
+            GT_FUNCTION_DEVICE auto warp_reduce(min, unsigned mask, int res) { return __reduce_min_sync(mask, res); }
+            GT_FUNCTION_DEVICE auto warp_reduce(max, unsigned mask, int res) { return __reduce_max_sync(mask, res); }
+            GT_FUNCTION_DEVICE auto warp_reduce(plus, unsigned mask, unsigned res) {
+                return __reduce_add_sync(mask, res);
+            }
+            GT_FUNCTION_DEVICE auto warp_reduce(min, unsigned mask, unsigned res) {
+                return __reduce_min_sync(mask, res);
+            }
+            GT_FUNCTION_DEVICE auto warp_reduce(max, unsigned mask, unsigned res) {
+                return __reduce_max_sync(mask, res);
+            }
+            GT_FUNCTION_DEVICE auto warp_reduce(bitwise_and, unsigned mask, unsigned res) {
+                return __reduce_and_sync(mask, res);
+            }
+            GT_FUNCTION_DEVICE auto warp_reduce(bitwise_or, unsigned mask, unsigned res) {
+                return __reduce_or_sync(mask, res);
+            }
+            GT_FUNCTION_DEVICE auto warp_reduce(bitwise_xor, unsigned mask, unsigned res) {
+                return __reduce_xor_sync(mask, res);
+            }
+#endif
+
+            template <size_t BlockSize, size_t ShmemSize = BlockSize / 32>
+            struct big_warp_kernel {
+                static_assert(BlockSize > 32, GT_INTERNAL_ERROR);
+                static_assert(BlockSize <= 1024, GT_INTERNAL_ERROR);
+                template <class F, class T>
+                GT_FUNCTION_DEVICE void operator()(F f, T const *__restrict__ in, T *__restrict__ out) const {
+                    __shared__ T buff[ShmemSize];
+                    in += blockIdx.x * BlockSize * 2 + threadIdx.x;
+                    T res = warp_reduce(f, 0xFFFFFFFF, f(*in, *(in + BlockSize)));
+                    if (threadIdx.x % 32 == 0)
+                        buff[threadIdx.x / 32] = res;
+                    __syncthreads();
+#ifdef __HIP__
+                    const unsigned ballot_result = __ballot(threadIdx.x < ShmemSize);
+#else
+                    const unsigned ballot_result = __ballot_sync(0xFFFFFFFF, threadIdx.x < ShmemSize);
+#endif
+                    if (threadIdx.x < ShmemSize)
+                        res = warp_reduce(f, ballot_result, buff[threadIdx.x]);
```

Comment:
Here, `warp_reduce` is called with a mask that is currently ignored in the HIP implementation.